cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -722,25 +736,60 @@ class ListConverter : public TypedConverterVisitor<ListBuilder, ListConverter> {
  public:
   Status Init(ArrayBuilder* builder) override;
 
-  Status AppendItem(const OwnedRef& item) {
+  Status AppendItem(PyObject* obj) override {
     RETURN_NOT_OK(typed_builder_->Append());
-    PyObject* item_obj = item.obj();
-    const auto list_size = static_cast<int64_t>(PySequence_Size(item_obj));
-    return value_converter_->AppendData(item_obj, list_size);
+    const auto list_size = static_cast<int64_t>(PySequence_Size(obj));
+    return value_converter_->AppendMultiple(obj, list_size);
   }
 
  protected:
   std::shared_ptr<SeqConverter> value_converter_;
 };
 
+class StructConverter : public TypedConverterVisitor<StructBuilder, StructConverter> {
+ public:
+  Status Init(ArrayBuilder* builder) override;
+
+  Status AppendItem(PyObject* obj) override {
+    RETURN_NOT_OK(typed_builder_->Append());
+    if (!PyDict_Check(obj)) {
+      return Status::TypeError("dict value expected for struct type");
+    }
+    // NOTE we're ignoring any extraneous dict items
+    for (int i = 0; i < num_fields_; i++) {
+      PyObject* nameobj = PyList_GET_ITEM(field_name_list_.obj(), i);
+      PyObject* valueobj = PyDict_GetItem(obj, nameobj);  // borrowed
```

Comment:
At some point we may set null if a particular key is not found, raising KeyError seems OK right now

Reply:
`PyDict_GetItem` doesn't raise KeyError, it simply returns NULL when the key isn't found (see also the tests which exercise this case).

Reply:
Noted, thank you =)